apps/prairielearn/src/ee/pages/lti13CourseNavigation/lti13CourseNavigation.ts

```diff
@@ -1,10 +1,182 @@
-router.all('/', (_req, res) => {
-  res.redirect('/pl');
-});
+router.get(
+  '/',
+  asyncHandler(async (req, res) => {
+    if ('done' in req.query) {
+      res.send(
+        Lti13CourseNavigationDone({
+          resLocals: res.locals,
+          lti13_instance_id: req.params.lti13_instance_id,
+        }),
+      );
+      return;
+    }
+
+    const ltiClaim = new Lti13Claim(req);
+    const courseName = `${ltiClaim.context?.label}: ${ltiClaim.context?.title}`;
+    const role_instructor = ltiClaim.isRoleInstructor();
+
+    // Get lti13_course_instance info, if present
+    const lti13_course_instance = await queryOptionalRow(
+      sql.select_course_instance,
+      {
+        lti13_instance_id: req.params.lti13_instance_id,
+        deployment_id: ltiClaim.deployment_id,
+        context_id: ltiClaim.context?.id,
+      },
+      Lti13CourseInstanceSchema,
+    );
+
+    if (lti13_course_instance) {
+      // Update lti13_course_instance on instructor login
+      // helpful as LMS updates or we add features
+      if (role_instructor) {
+        await queryAsync(sql.update_lti13_course_instance, {
+          lti13_instance_id: req.params.lti13_instance_id,
+          course_instance_id: lti13_course_instance.course_instance_id,
+          deployment_id: ltiClaim.deployment_id,
+          context_id: ltiClaim.context?.id,
+          context_label: ltiClaim.context?.label,
+          context_title: ltiClaim.context?.title,
+        });
+
+        // TODO: Set course/instance staff permissions for LMS course staff here?
+      }
+
+      // LTI claims are not used after this page so remove them from the session
+      ltiClaim.remove();
+
+      // Redirect to linked course instance
+      res.redirect(
+        `/pl/course_instance/${lti13_course_instance.course_instance_id}/${
+          role_instructor ? 'instructor/' : ''
+        }`,
+      );
+      return;
+    }
+
+    // Students get a "come back later" message
+    if (!role_instructor) {
+      res.send(
+        Lti13CourseNavigationNotReady({
+          resLocals: res.locals,
+          courseName,
+        }),
+      );
+      return;
+    }
+
+    // Instructor so lookup their existing information in PL
+    const courses = await selectCoursesWithEditAccess({
+      user_id: res.locals.authn_user.user_id,
+      is_administrator: res.locals.authn_is_administrator,
+    });
+
+    const course_instances: CourseInstanceAuthz[] = [];
+
+    for (const course of courses) {
+      // Only course owners can link
+      if (!course.permissions_course.has_course_permission_own) {
+        continue;
+      }
+
+      const instances = await selectCourseInstancesWithStaffAccess({
+        course_id: course.id,
+        user_id: res.locals.authn_user.user_id,
+        authn_user_id: res.locals.authn_user.user_id,
+        is_administrator: res.locals.authn_is_administrator,
+        authn_is_administrator: res.locals.authn_is_administrator,
+      });
```

Comment:
This is going to be a horrible N+1 query for any admins. Two options:

- Ignore it and pray that admins don't ever try to initialize LTI flows. With impending institution admins, we could actually see this happening in production.
- Ignore it and assume that a) Postgres will be fast enough and b) this will happen very infrequently.
- Turn this into a multi-step flow where the user needs to first select a course, at which point we show them a list of course instances only for that course. Two possible ways we could do this:
  - Split the flow into multiple pages.
  - Keep the flow on one page but use htmx to dynamically load the list of course instances once a course is selected. We do something similar on the global admin feature grant modal: https://github.com/PrairieLearn/PrairieLearn/blob/fc1f40c7438bb3d28cc98e28c193ca59ad70c751/apps/prairielearn/src/pages/administratorFeatures/administratorFeatures.html.ts#L248

Reply:
I like the idea of htmx for course instance lookup based on course. I'll work on that after everything else in this PR is squashed.

Reply:
I also think that could be a follow on PR that fixes the logic. I think your first assumptions are correct: this won't typically happen (or at least often) and hopefully the query is fast enough that it doesn't bog anything down.

Reply:
I'm fine with this being done in a followup. Maybe just leave a comment acknowledging the N+1 issue and stating our rationale for why we're ok with it for now?

Reply:
Addressed in 10bfa667d